### PR TITLE
WIP: Specify file resolution and exports

### PR DESCRIPTION
**Pull request by guybedford**
This PR combines the spec for the file resolution proposal in https://github.com/nodejs/ecmascript-modules/pull/19, with the exports spec from https://github.com/jkrems/proposal-pkg-exports/.

I've put together an experimental implementation of this spec at https://github.com/guybedford/node/tree/irp-implementation if anyone wants to try it out.

Feedback welcome.

### CODE REVIEW

File: doc/api/esm.md [hunk 1]
```diff
@@ -157,37 +149,177 @@ The resolver has the following properties:
...
+>    1. Return **false**.
+> 1. If _target_ contains any percent-encoded characters for _"/"_ or _"\"_,
+>    then
+>    1. Return **false**.
+> 1. Return **true**.
+
+**ESM_FORMAT(_url_)**
+> 1. Assert: _url_ corresponds to an existing file.
+> 1. Let _pjson_ be the result of **READ_PACKAGE_BOUNDARY**(_url_).
+> 1. If **HAS_ESM_PROPERTIES**(_pjson_) is **true**, then
```
Comment (devsnek) [inline doc/api/esm.md]: this still needs to be fixed. the default behaviour should always be to treat mjs as esm

Comment (guybedford) [reply doc/api/esm.md]: The implementation actually does this - it was a spec error which is resolved now.

File: doc/api/esm.md [hunk 2]
```diff
@@ -157,37 +149,187 @@ The resolver has the following properties:
...
+> 1. Return **null**.
+
+READ_JSON_FILE(_url_)
+> 1. If the file at _url_ does not parse as valid JSON, then
+>    1. Throw an _Invalid Package Configuration_ error.
+> 1. Let _pjson_ be the parsed JSON source of the file at _url_.
+> 1. Return _pjson_.
+
+HAS_ESM_PROPERTIES(_pjson_)
+> 1. If _pjson_ is not **null** and _pjson.exports_ is a String or Object, then
```
[the remark this answers is not in the document]
Comment (devsnek) [reply doc/api/esm.md]: It's also worth noting that this has a significant impact on how node deals with package.json from a performance standpoint.

**Comment by MylesBorins** [comment]
@guybedford gone ahead and rebase this against the new modules-lkgr

**Comment by guybedford** [comment]
Replaced by #28.

**Pull request closed without merging**